classes/class.tapestry.php

```diff
@@ -526,30 +526,34 @@ private function _filterNodeMetaIdsByPermissions($nodeMetaIds, $rootId, $superus
 
     private function _pathIsAllowed($from, $to, $checked = [], $superuser_override, $userId)
     {
-        if (TapestryCache::exists(__METHOD__, [$from, $to, $superuser_override, $userId])) {
-            return TapestryCache::get(__METHOD__, [$from, $to, $superuser_override, $userId]);
-        }
-
-        $pathIsAllowed = false;
-        if (!in_array($from, $checked) && (
-            TapestryHelpers::userIsAllowed('READ', $from, $this->postId, $superuser_override, $userId)
-            || (TapestryHelpers::userIsAllowed('ADD', $from, $this->postId, $superuser_override, $userId)
-            || TapestryHelpers::userIsAllowed('EDIT', $from, $this->postId, $superuser_override, $userId))
-        )) {
-            if ($from == $to) {
-                $pathIsAllowed = true;
-            } else {
-                $checked[] = $from;
+        $queue = new SplQueue();
+        $checked[] = $from;
+        $queue->enqueue($from);
+        while (!$queue->isEmpty()) {
+            $node = $queue->dequeue();
+            if (TapestryCache::get(__METHOD__, [$node, $to, $superuser_override, $userId])) {
```

Comment:
If you are just checking if something is cached, you should use the `exists` method, not the `get` method.

Reply:
Fair. This change is probably slowing it down because I'm only getting the cached value if it's true. I was coming across issues with a node without read permissions, caching as false and then a subsequent adjacent node retrieves that and thinks it is also false.
![image](https://user-images.githubusercontent.com/45089842/95155862-f509cf80-0749-11eb-8f3e-512ec67671b1.png)
Will come back with a solution...